akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private trait PhaseDefinition {
+    def size: Int
+    def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+  }
+
+  private object tasks {
+    private val registeredPhases = new ConcurrentHashMap[String, StrictPhaseDefinition]()
+    private trait TaskDefinition extends Cancellable {
+      private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done]
+    }
+    private object TaskDefinition {
+      def create(phaseName: String, task: () => Future[Done], name: String): TaskDefinition = new TaskDefinition {
+        // This is a vanilla class instead of a case class to avoid default implementations of .hashCode() and .equals
+        // Different code paths could register the same task under the same name multiple times; in this case we want to
+        // run that task as many times as it was registered (minus the number of times those were cancelled), so they must
+        // be distinct in a Set[TaskDefinition].
+
+        private val p = Promise[Done]()
+
+        override private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done] = {
+          if (log.isDebugEnabled) {
+            log.debug("Performing task [{}] in CoordinatedShutdown phase [{}]", name, phaseName)
+          }
+          if (p.isCompleted) {
+            Future.successful(Done)
```

Comment:
Shouldn't this be considered a bug, even that `run` was invoked twice rather than that the promise completed perhaps? The user can submit the same `task` several times, but the `TaskDefinition` should only ever be added once, right?

Reply:
It is probably a bug if `run` ever gets invoked twice, but depending on when a task gets pruned from the tasks hashmap, it is conceivable that `run` could be called on a cancelled task, which shouldn't complain (while also not running the task).